Add Streaming support

millermedeiros opened the pull request:
I ended up using 2 different APIs (one for the `client` and another one for the `service`).

---

closes #6

millermedeiros commented:
## Client API

```js
var stream = client.stream('contacts', { some: 'data' });
stream.on('data', onData);
stream.on('error', onError);
stream.on('close', onClose);
stream.abort('because I want').then(doStuff, onError);
```

## Service API

```js
threads.service('my-service')
  .stream('contacts', (stream, data) => {
    stream.onabort = cancelEverything;
    if (data.foo === 123) return stream.error('invalid argument');
    stream.write('stuff');
    stream.close();
  });
```

wilsonpage commented on lib/client-stream.js:
```diff
@@ -0,0 +1,27 @@
+}
+
+ClientStream.prototype = {
+  __proto__: Emitter.prototype,
```
Is this smelly? Can we use `Object.create()`.

millermedeiros replied:
sure

wilsonpage commented on lib/service.js:
```diff
@@ -114,6 +122,8 @@ function ServicePrivate(service, name) {
   this.methods = {};
   this.channels = {};
+  this.streams = {};
+  this._activeStreams = {};
```
NIT: `_` not required as this is inside the 'private' class.

millermedeiros replied:
sure

wilsonpage commented on lib/service-stream.js:
```diff
@@ -0,0 +1,54 @@
+    this._validateState('abort');
+    this._state = 'aborted';
+    if (!('onabort' in this)) {
+      throw new Error('stream.onabort() not implemented by service.');
```
What if the service isn't interested in this event?

wilsonpage replied:
An reason why we're not using `Emitter` here?

wilsonpage commented on lib/service.js:
```diff
@@ -75,6 +78,11 @@ Service.prototype.method = function(name, fn) {
 };
 
+Service.prototype.stream = function(name, fn) {
+  this.private.streams[name] = fn;
```
Let's make an `this.private.addStream()` to prevent touching internals.

wilsonpage commented on lib/service-stream.js:
```diff
@@ -0,0 +1,54 @@
+
+module.exports = ServiceStream;
+
+function ServiceStream(service, id) {
```
I feel like it might make sense to pass the client's `channel` into the `ServiceStream` so that we can pipe messages down it. Right now it looks like we're broadcasting stream messages to all clients. This might mean we can kill the new `._broadcast` API which is confusing me.

wilsonpage replied:
A stream will only ever have one client right?

millermedeiros replied:
yes, a stream should only have a single client.

wilsonpage commented on lib/service.js:
```diff
@@ -360,13 +396,17 @@ ServicePrivate.prototype.listen = function() {
+  });
+};
+
+ServicePrivate.prototype._broadcast = function(messageId, data) {
```
As mentioned above: If we passed the `this.channels[client]` into the `ServiceStream` it could send messages directly to the client on it's own.

wilsonpage commented on lib/client.js:
```diff
@@ -5,6 +5,7 @@
  * Dependencies
  */
 
```
We should decide on a 'private' pattern (like in `Service`) and use it across all our classes.

millermedeiros replied:
on the streams I used `this._ = PrivateStream()` since it's easy to type and it's common to have `_` as the start of private members..

 `this._.foo` also looks cool because `._.` is almost an emoji :D

millermedeiros replied:
exposing the private methods/properties is good because it allows us to unit test, see: https://github.com/millermedeiros/threads/commit/843425d5e0a2a64f2eb1794bf7e0eefd3758af5c#diff-3d6cd7d35d59ce33f7b9c33ff2d59a20R20

wilsonpage replied:
:heart: ._.

wilsonpage commented on lib/client.js:
```diff
@@ -242,6 +247,47 @@ Client.prototype.method = function(method) {
+  return stream;
+};
+
+Client.prototype.onstreamdata = function(broadcast) {
```
Might be simpler to remove these three events and replace them with:

``` js
Client.prototype.onstreamevent = function(event) {
  this._activeStreams[event.stream].emit(event.type, event.data);
};
```

Then if you need to do any internal house-keeping on any particular events you can bind to their public API just after they're created.

``` js
stream
  .on('close', deleteStream);
  .on('error', deleteStream);
```

millermedeiros commented:
I just updated the PR with a new API, implementation is still not as clean as it could be but I think the new API is closer to the [WHATWG streams](https://streams.spec.whatwg.org/)

## Client API

```js
var stream = client.stream('contacts', { some: 'data' });
stream.listen(onData);
stream.closed.then(onClose, onAbort);
stream.cancel('because I want').then(doStuff, onError);
```

## Service API

```js
threads.service('my-service')
  .stream('contacts', (stream, data) => {
    stream.oncancel = cancelEverything;
    if (data.foo === 123) return stream.abort('invalid argument');
    stream.write('stuff');
    stream.close();
  });
```

millermedeiros commented on threads.js:
```diff
@@ -164,7 +164,63 @@ ChildThread.prototype.destroyProcess = function() {
+    return {
+      cancel: function() {
+        emitter.off('data', callback);
+      }
```
uhm.. now that I'm thinking about it, it would be *better* to have an `unlisten` or `removeListener` method instead of returning a new object.

millermedeiros commented on lib/client-stream.js:
```diff
@@ -0,0 +1,54 @@
+      cancel: function() {
+        emitter.off('data', callback);
+      }
+    };
```
uhm.. now that I'm thinking about it, it would be _better_ to have an `unlisten` or `removeListener` method instead of returning a new object.

millermedeiros commented:
# API v3

closer to the [WHATWG streams](https://streams.spec.whatwg.org/) and simpler to use.

## Client API

```js
var stream = client.stream('contacts', { some: 'data' });
stream.listen(onData);
stream.unlisten(onData);
stream.closed.then(onClose, onAbort);
stream.cancel('because I want').then(doStuffAfterCancel).catch(onCancelError);
```

## Service API

```js
threads.service('my-service')
  .stream('contacts', (stream, data) => {
    stream.cancel = cancelEverything;
    if (data.foo === 123) return stream.abort('invalid argument');
    stream.write('stuff');
    stream.close();
  });
```

and now instead of throwing errors I'm returning rejected promises on `write`, `close` and `abort` (in case service tries to call `write` after `abort/close` or call `abort/close` multiple times in a row) kinda following the whatwg writable stream behavior. - we ignore the action without throwing an error since we assume that after `abort`/`close` the stream is going to be destroyed and also because throwing errors on async operations is an anti-pattern.

millermedeiros commented:
I believe the patch is ready for review, added unit tests for the edge cases, JSDoc comments for public APIs, refactored... I'll squash the commits after I get r+

wilsonpage commented on lib/service.js:
```diff
@@ -7,6 +7,7 @@
 
 var thread = require('./thread-global');
 var utils = require('./utils');
+var ServiceStream = require('./service-stream');
```
Move this to line 8 to satisfy my irrational OCD-ness :stuck_out_tongue_winking_eye:

wilsonpage commented on lib/service.js:
```diff
@@ -360,13 +407,17 @@ ServicePrivate.prototype.listen = function() {
+  });
+};
+
+ServicePrivate.prototype.postMessage = function(messageId, data) {
```
Can we get some jsdocs on all private function too? Even though they're not public it helps me follow the code :)

wilsonpage commented on test/service-stream.test.js:
```diff
@@ -0,0 +1,190 @@
+
+    function callStream(methodName, fn) {
+      service.stream(methodName, fn);
+      service.private.channels['123'] = {
```
Would it be possible to test these edge cases without tying implementation details to the tests?

millermedeiros replied:
it would if I could access the `ServiceStream` constructor -- which is a private constructor at the moment and our test infrastructure don't really load the src files, just the built file.

wilsonpage commented on lib/client.js:
```diff
@@ -242,6 +245,37 @@ Client.prototype.method = function(method) {
   });
 };
 
+Client.prototype.stream = function(method) {
```
Docs needed

wilsonpage commented on threads.js:
```diff
@@ -731,15 +995,56 @@ ServiceInternal.prototype.onrequest = function(request) {
+  var id = method.id;
+  var client = request.client;
+  var channel = this.channels[client];
+  var stream = new ServiceStream(id, channel, this.id, client);
```
Number of args passsed to `ServiceStream` is getting lengthy. Perhaps a `params` object might be more readable. What do you think?

wilsonpage commented on threads.js:
```diff
@@ -575,7 +704,106 @@ ManagerInternal.prototype.createThread = function(descriptor) {
+  this.message = new utils.Messages(this, serviceId, []);
+}
+
+PrivateServiceStream.prototype.validateState = function(actionName, state) {
```
Why doesn't this function return a promise? Can't see any async stuff.

millermedeiros replied:
it returns a promise because `cancel`, `abort`, `close` and `write` should only really send data to the client if it's on a valid state and they should return rejected promises when they fail.. so it made sense to do `return this.validateState('cancel', 'canceled').then(...)` but I'll replace the logic since it's confusing.

wilsonpage commented:
Just a few NITs really. Mainly meed jsdoc coverage to make code easier to understand. Overall this is looking really quality! Your code is slick :)

millermedeiros committed:
Add streaming support. closes #6

millermedeiros commented:
squashed commits and exposed the `ServiceStream` as `threads.service.Stream`
